proposals/0000-linear-types.rst

```diff
@@ -0,0 +1,2245 @@
+- ``u = λy. v`` and the usage of ``x`` in ``v`` is ``p``.
+
+A variable's usage is correct if it is smaller than or equal to the
+multiplicity annotation of the variable (note that 0 is *not* smaller
+than one). Incorrect usage results in a type error. This definition is
+close to the intended implementation of multiplicities. The `paper
+<https://arxiv.org/abs/1710.09756>`_ has a more declarative
+definition.
+
+The multiplicity of a variable introduced by a λ-abstraction is taken
+from the surrounding typing information (typically a type annotation
+on an equation). For instance
+
+::
+
+  foo :: A :p-> B
+  foo x = …  -- x has multiplicity p
+
+The above takes care of the pure λ-calculus part of Haskell. We also
+need to consider ``let`` and ``case``.
+
+A ``let`` binding is considered to have an implicit multiplicity
+annotation (the annotation is inferred). The variables introduced by a
+``let`` bindings with annotation ``p`` all have multiplicity
+``p``. And the usage of ``x`` in ``let_p {y1 = u1; … ;yn = un} in v``
+(where the ``yi`` are variables) is ``p*q1 + … + p*qn + q`` where the
+usage of ``x`` in ``ui`` is ``qi`` and in ``v`` is ``q``.
+
+If a let has recursive binders, then ``p`` must be ``ω``.
+
+A ``case`` expression has an implicit multiplicity annotation, like
+``let`` binding. It if often inferred from the type annotation of an
+equation. The usage of ``x`` in ``case_p u of { … }``, where the usage
+of ``x`` in ``u`` is ``q`` is ``p*q`` plus the *join* of the usage of
+``x`` in each branch.  Note that, in usages, ``0 ≰ 1`` as arguments
+with multiplicity ``1`` are consumed exactly once, which doesn't
+include not being consumed at all.
+
+The multiplicity annotation of variables introduced by a pattern depend
+on the constructor and on the implicit annotation of the
+``case``. Specifically in ``case_p u of {…; C x1 … xn -> …; …}`` Where ``C :: a1 :q1-> … an :qn-> A``,
+Then ``xi`` has multiplicity annotation ``p*qi``. For instance
+
+::
+
+  bar :: (a,b) :p-> c
+  bar (x,y) = … -- Since (,) :: a ->. b ->. (a,b), x and y have
+                -- multiplicity p
+
+Deep patterns & multiple-argument equations
+-------------------------------------------
+
+.. _Patterns:
+
+Type-checking deep patterns naturally extends the simple patterns
+above. For instance in
+
+::
+
+  f :: Maybe (a, b) ->. …
+  f (Just (x,y)) = …
+
+since the type annotation on the first argument is linear, the outer
+pattern is type-checked as a ``case_1``:
+
+::
+
+  f mxy = case_1 mxy of
+    Just xy -> …
+
+Therefore, the generated intermediate variable ``xy`` has multiplicity
+1, therefore, the inner pattern is elaborated as a ``case_1`` (that is
+the same multiplicity as the intermediate variable).
+
+::
+
+  f mxy = case_1 mxy of
+    Just xy -> case_1 xy of
+      (x, y) -> …
+
+Typechecking equations
+~~~~~~~~~~~~~~~~~~~~~~
+
+In a definition with multiple equations, each equation is typechecked
```

Comment:
21. This is another place where the "consumed exactly once" definition led me astray. A failed pattern match can do an arbitrary amount of work and consumption before failing -- it just can't record any results of that work.